python/ray/_private/package.py

```diff
@@ -0,0 +1,160 @@
+import hashlib
+import logging
+import inspect
+
+from filelock import FileLock
+from pathlib import Path
+from zipfile import ZipFile
+from ray.job_config import JobConfig
+from enum import Enum
+from ray.experimental import internal_kv
+from urllib.parse import urlparse, urlunparse
+from ray.core.generated import gcs_pb2
+import os
+import sys
+
+PKG_DIR="/tmp/ray/runtime_resources"
+Path(PKG_DIR).mkdir(parents=True, exist_ok=True)
+
+logger = logging.getLogger(__name__)
+
+class Protocol(Enum):
+    # gcs is for package created by system and will be managed by
+    # ray cluster automatically through reference counting
+    GCS = "gcs"
+    # pin_gcs is for package created mannually. User needs to
+    # maintain the lifetime of this object
+    PIN_GCS = "pingcs"
+
+def _xor_bytes(left, right):
+    if left and right:
+        return bytes(a ^ b for (a, b) in zip(left, right))
+    return left or right
+
+def _zip_module(path, relative_path, zip_handler):
+    for from_file_name in path.glob("**/*"):
+        to_file_name = from_file_name.relative_to(relative_path)
+        zip_handler.write(from_file_name, to_file_name)
+
+def _hash_modules(path):
+    hash_val = None
+    BUF_SIZE = 4096 * 1024
+    for from_file_name in path.glob("**/*"):
+        md5 = hashlib.md5()
+        if Path(from_file_name).is_dir():
+            md5.update(str(from_file_name).encode())
+        else:
+            with open(from_file_name, mode="rb") as f:
+                data = f.read(BUF_SIZE)
+                if not data:
+                    break
+                md5.update(data)
+        hash_val = _xor_bytes(hash_val, md5.digest())
+    return hash_val
+
+def get_local_path(pkg_uri) -> str:
+    (_, pkg_name) = parse_uri(pkg_uri)
+    return os.path.join(PKG_DIR, pkg_name)
+
+# TODO(yic): Fix this later to handle big directories in better way
+def get_project_package_name(working_dir, modules) -> str:
```

Comment:
Please add pydoc for this function. You can see examples in other python files in ray.